### PR TITLE
Issue #274: Avoid premature vault token renewals

### PR DESCRIPTION
**Pull request by pschultz**
Don't attempt Vault token renewals if the token isn't renewable. If it
is, don't renew the token with each refresh; only do so if the token
would expire shortly. Increase the token's lifetime by its original TTL.

This now requires the "read" capability for the "auth/token/lookup-self" path in Vault. That is granted by default to all tokens. If the lookup-self request fails for any reason, fabio will assume that the token is not renewable.

No new configuration is required. Automatic renewal will work as long as the refresh interval is smaller than the token TTL, same as before.

### CODE REVIEW

**Comment by pschultz** [comment]
The builds failed because of network issues between Travis and HashiCorp:

> $ wget https://releases.hashicorp.com/consul/0.8.3/consul_0.8.3_linux_amd64.zip
> --2017-06-28 14:01:50--  https://releases.hashicorp.com/consul/0.8.3/consul_0.8.3_linux_amd64.zip
> Resolving releases.hashicorp.com (releases.hashicorp.com)... 151.101.89.183, 2a04:4e42:11::439
> Connecting to releases.hashicorp.com (releases.hashicorp.com)|151.101.89.183|:443... failed: Connection timed out.
> Connecting to releases.hashicorp.com (releases.hashicorp.com)|2a04:4e42:11::439|:443... failed: Network is unreachable.

I don't think I can retry them.

File: cert/vault_source.go [hunk 1]
```diff
@@ -45,20 +50,27 @@ func (s *VaultSource) client() (*api.Client, error) {
...
+	if err := s.setAuth(c); err != nil {
 		return nil, err
 	}
+
 	return c, nil
 }
 
-func (s *VaultSource) setToken(c *api.Client) error {
+func (s *VaultSource) setAuth(c *api.Client) error {
+	firstCall := true
```
Comment (magiconair) [inline cert/vault_source.go]: This looks racy and should probably be inside the lock.

File: cert/vault_source.go [hunk 2]
```diff
@@ -71,17 +83,53 @@ func (s *VaultSource) setToken(c *api.Client) error {
...
 
+// dropNotRenewableWarning controls whether the 'Token is not renewable'
+// warning is logged. This is useful for testing where this is the expected
+// behavior. On production, this should always be set to false.
+var dropNotRenewableWarning bool
+
+// checkRenewal checks if the Vault token can be renewed, and if so when it
+// expires and how big the renewal increment should be.
+func (s *VaultSource) checkRenewal(c *api.Client) {
+	response, err := c.Auth().Token().LookupSelf()
```
Comment (magiconair) [inline cert/vault_source.go]: `s/response/resp/`

File: cert/vault_source.go [hunk 3]
```diff
@@ -71,17 +83,53 @@ func (s *VaultSource) setToken(c *api.Client) error {
...
+		CreationTTL int       `json:"creation_ttl"`
+		ExpireTime  time.Time `json:"expire_time"`
+		Renewable   bool      `json:"renewable"`
+	}
+	if err := json.Unmarshal(b, &data); err != nil {
+		log.Printf("[WARN] vault: lookup-self failed, token renewal is disabled: %s", err)
+		return
+	}
+
+	if data.Renewable {
```
Comment (magiconair) [inline cert/vault_source.go]: pls rewrite as `switch` statement. I try to avoid if/else cascades.

File: cert/vault_source.go [hunk 4]
```diff
@@ -165,3 +201,37 @@ func (s *VaultSource) load(path string) (pemBlocks map[string][]byte, err error)
...
+		// if s.Refresh is small, say one second. It is risky to renew the
+		// token just one or two seconds before expiration; networks are
+		// unreliable, clocks can be skewed, etc.
+		break
+	default:
+		// Token doesn't need to be renewed yet.
+		return nil
+	}
+
+	response, err := c.Auth().Token().RenewSelf(s.auth.renewIncrement)
```
Comment (magiconair) [inline cert/vault_source.go]: `s/response/resp/`

File: cert/vault_source.go [hunk 5]
```diff
@@ -165,3 +201,37 @@ func (s *VaultSource) load(path string) (pemBlocks map[string][]byte, err error)
...
+
+	ttl := time.Until(s.auth.expireTime)
+	switch {
+	case s.auth.expireTime.IsZero():
+		// Token isn't renewable.
+		return nil
+	case ttl < 2*s.Refresh:
+		// Renew the token if it isn't valid for two more refresh intervals.
+		break
+	case ttl < 1*time.Minute:
```
Comment (magiconair) [inline cert/vault_source.go]: pls drop the `1*` since it is redundant.

File: cert/vault_source.go [hunk 6]
```diff
@@ -29,8 +30,12 @@ type VaultSource struct {
 	Refresh      time.Duration
 
 	mu         sync.Mutex
-	token      string // actual token
 	vaultToken string // VAULT_TOKEN env var. Might be wrapped.
+	auth       struct {
+		token          string    // actual token
+		expireTime     time.Time // zero value if the token isn't renewable
+		renewIncrement int
```
Comment (magiconair) [inline cert/vault_source.go]: pls add a comment and explain the unit. Maybe a shorter name, e.g. `renewTTL`

File: cert/vault_source.go (hunk 5 above)
Comment (pschultz) [inline cert/vault_source.go]: It is redundant, but it reads as "one minute" which I prefer. This is a constant expression that will be evaluated at compile time, so there is no runtime impact. If you insist I'll change it though.

File: cert/vault_source.go (hunk 1 above)
Comment (pschultz) [inline cert/vault_source.go]: Technically not a race (all reads and writes are inside the lock), but I moved it down anyway to remove any doubts.

**Comment by pschultz** [comment]
Please take another look. I will fixup into a single commit once you're good with the PR.

File: cert/vault_source.go [hunk 7]
```diff
@@ -32,9 +32,12 @@ type VaultSource struct {
 	mu         sync.Mutex
 	vaultToken string // VAULT_TOKEN env var. Might be wrapped.
 	auth       struct {
-		token          string    // actual token
-		expireTime     time.Time // zero value if the token isn't renewable
-		renewIncrement int
+		token      string    // actual token
```
Comment (magiconair) [inline cert/vault_source.go]: pls use the Go doc standard here:

```
// token is the actual token
token string

// expireTime is ...
expireTime time.Time
...
```

File: cert/vault_source.go [hunk 8]
```diff
@@ -119,15 +124,21 @@ func (s *VaultSource) checkRenewal(c *api.Client) {
...
+	case data.ExpireTime.IsZero():
+		// token doesn't expire
+		return
+	case dropNotRenewableWarning:
+		return
 	}
+
+	ttl := time.Until(data.ExpireTime)
+	ttl = ttl / time.Second * time.Second // truncate to seconds
+	log.Printf("[WARN] vault: Token is not renewable and will expire %s from now (at %s)",
```
Comment (magiconair) [inline cert/vault_source.go]: The logging is the `default` case, right?

pls change the log msg to `Token is not renewable and expires in %s from now at %s`

**Comment by magiconair** [comment]
Don't bother. I can squash the merge myself.

File: cert/vault_source.go (hunk 5 above)
Comment (magiconair) [inline cert/vault_source.go]: I don't insist but I'd prefer since it is redundant :)

File: cert/vault_source.go (hunk 1 above)
Comment (magiconair) [inline cert/vault_source.go]: Right, since this is a local var but then this doesn't make sense. The `defer` function should then read:

```go
defer func() {
    c.SetToken(s.auth.token)
    if s.auth.token != "" {
        s.checkRenewal(c)
    }
    s.mu.Unlock()
}
```

**Comment by magiconair** [comment]
@pschultz Cheeky side question: Would you mind me adding your logo to the README? :)

File: cert/vault_source.go (hunk 1 above)
Comment (pschultz) [inline cert/vault_source.go]: s.auth.token will never be empty when the deferred function executes (assuming there is no error). Apparently I made this is too confusing. I think we can use sync.Once to make it obvious. Let my try that.

Comment (magiconair) [inline cert/vault_source.go]: I also suggest to remove the `SetToken` call from the `defer` method and put it where necessary. There is too much magic going on here which is hard to follow.

Comment (pschultz) [inline cert/vault_source.go]: See 49767dd

**Comment by pschultz** [comment]
Sure, you can add the logo. We don't have an English site you can link to though, only the one in German.

**Comment by magiconair** [comment]
I think this is more readable. What do you think?

```go
func (s *VaultSource) setAuth(c *api.Client) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.auth.token != "" {
		return nil
	}

	if s.vaultToken == "" {
		return errors.New("vault: no token")
	}

	// did we get a wrapped token?
	resp, err := c.Logical().Unwrap(s.vaultToken)
	if err != nil {
		// unwrapping failed?
		if !strings..HasPrefix(err.Error(), "no value found at") {
			return err
		}
		s.auth.token = s.vaultToken
	} else {
		log.Printf("[INFO] vault: Unwrapped token %s", s.vaultToken)
		s.auth.token = resp.Auth.ClientToken
	}

	s.auth.once.Do(func() { s.checkRenewal(c) })
	c.SetToken(s.auth.token)
	return nil
}
```

**Comment by pschultz** [comment]
Yeah, that's better. The last if/else statements can even be done with a nice switch. See f6e2463.

**Comment by magiconair** [comment]
LGTM
